lib/backend/migration/migration.go

```diff
@@ -0,0 +1,155 @@
+package migration
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/trace"
+)
+
+// Migration manages a migration between two [backend.Backend] interfaces.
+type Migration struct {
+	src      backend.Backend
+	dst      backend.Backend
+	parallel int
+	total    int
+	migrated atomic.Int64
+	log      logrus.FieldLogger
+}
+
+// MigrationConfig configures a [Migration] with a source and destination backend.
+// All items from the source are copied to the destination. All Teleport Auth
+// Service instances should be stopped when running a migration to avoid data
+// inconsistencies.
+type MigrationConfig struct {
+	// Source is the backend [backend.Config] items are migrated from.
+	Source backend.Config `yaml:"src"`
+	// Destination is the [backend.Config] items are migrated to.
+	Destination backend.Config `yaml:"dst"`
+	// Parallel is the number of items that will be migraated in parallel.
+	Parallel int `yaml:"parallel"`
+	// Log logs the progress of a [Migration]
+	Log logrus.FieldLogger
+}
+
+// New returns a [Migration] based on the provided [MigrationConfig].
+func New(ctx context.Context, config MigrationConfig) (*Migration, error) {
+	src, err := backend.New(ctx, config.Source.Type, config.Source.Params)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create source backend")
+	}
+	dst, err := backend.New(ctx, config.Destination.Type, config.Destination.Params)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create destination backend")
+	}
+	migration := &Migration{
+		src:      src,
+		dst:      dst,
+		parallel: config.Parallel,
+		log:      config.Log,
+	}
+	if migration.parallel == 0 {
+		migration.parallel = 1
+	}
+	if migration.log == nil {
+		migration.log = logrus.WithField(teleport.ComponentKey, "migration")
+	}
+	return migration, nil
+}
+
+// Run runs a [Migration] until complete.
+func (m *Migration) Run(ctx context.Context) error {
+	var all []backend.Item
+	start := backend.Key("")
```

Comment:
Some items are not actually between `/` and `0` - I believe it should only be locks tho, so perhaps that's even beneficial.